core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -250,6 +252,19 @@ protected static int getTotalAttempts(ClientRequestContext ctx) {
         return state.totalAttemptNo;
     }
 
+    /**
+     * Creates a new derived {@link ClientRequestContext}, replacing the {@link Request} with {@code req}.
+     * If {@link ClientRequestContext#endpointSelector()} exists, a new {@link Endpoint} will be selected.
+     */
+    protected static ClientRequestContext newDerivedContext(ClientRequestContext ctx, Request req) {
+        final EndpointSelector endpointSelector = ctx.endpointSelector();
+        if (endpointSelector != null) {
+            return ctx.newDerivedContext(req, endpointSelector.select(ctx));
+        } else {
+            return ctx.newDerivedContext(req);
```

Comment:
If the resolution is faied, this will throw `IllegalStateExeption` inside `newDerivcedCotext`.
Don't we need to throw `EndpointGroupException` for that?